traitsui/qt4/video_editor.py

```diff
@@ -0,0 +1,362 @@
+    state = PlayerState()
+
+    #: The current playback position of the player, synchronized to the trait
+    #: named by factory.position.
+    position = Float()
+
+    #: The total duration of the current video, synchronized to the trait
+    #: named by factory.duration.
+    duration = Float()
+
+    #: The media player playback status (loading, buffering, etc.).
+    media_status = MediaStatus()
+
+    #: The percentage of the buffer currently filled.
+    buffer = Range(0, 100)
+
+    #: A string holding the video error state, or "" if no error.
+    video_error = Str()
+
+    #: Whether the audio is muted or not.
+    muted = Bool(False)
+
+    #: The playback volume on a logarithmic scale (perceptually linear).
+    volume = Range(0.0, 100.0)
+
+    #: The playback rate.  Negative values rewind the video.
+    playback_rate = Float(1.0)
+
+    def _get_frame(self):
+        return self.control._np_image
+
+    def update_to_regular(self):
+        self.control = QVideoWidget()
+        self.control.setSizePolicy(
+            QSizePolicy.Expanding, QSizePolicy.Expanding
+        )
+        self.control.setBackgroundRole(QPalette.Window)
+        self.media_player.setVideoOutput(self.control)
+
+    def update_to_functional(self):
+        self.control = ImageWidget(image_fun=self.image_fun)
+        self.control.setSizePolicy(
+            QSizePolicy.Expanding, QSizePolicy.Expanding
+        )
+        self.control.setBackgroundRole(QPalette.Window)
+
+        self.surface = VideoSurface(widget=self.control)
+        self.surface.frameAvailable.connect(self.control.setImage)
+
+        self.media_player.setVideoOutput(self.surface)
+
+    def _image_fun_changed(self):
+        if self.image_fun is None:
+            self.update_to_regular()
+        else:
+            self.update_to_functional()
+
+    # ------------------------------------------------------------------------
+    # Editor interface
+    # ------------------------------------------------------------------------
+
+    def init(self, parent):
+        """Initialize the editor by creating the underlying toolkit widget.
+
+        Parameters
+        ----------
+        parent : QWidget or None
+            The parent widget for this widget.
+        """
+        self.control = QVideoWidget()
+        self.control.setSizePolicy(
+            QSizePolicy.Expanding, QSizePolicy.Expanding
+        )
+        self.control.setBackgroundRole(QPalette.Window)
+
+        self.media_player = QMediaPlayer(None, QMediaPlayer.VideoSurface)
+        self._set_video_url()
+        self.media_player.setVideoOutput(self.control)
+        self.media_player.setMuted(self.muted)
+        self._state_changed()
+        self._aspect_ratio_changed()
+        self._muted_changed()
+        self._volume_changed()
+        self._playback_rate_changed()
+
+        self._connect_signals()
+
+        self.set_tooltip()
+
+    def dispose(self):
+        self._disconnect_signals()
```

Comment:
This should call `super().dispose()`, no?

Reply:
Yes, almost certainly.